data_pipeline_api/registry/downloader.py

```diff
@@ -0,0 +1,372 @@
+                MetadataKey.extension: Path(block["output_filename"]).suffix[1:],
+                MetadataKey.filename: block["output_filename"],
+                MetadataKey.component: block[DataRegistryTarget.object_component, DataRegistryField.name],
+            }
+            metadatas.append(metadata)
+        if metadatas:
+            yaml.safe_dump(metadatas, stream)
+
+    def _write_metadata_external_object(self, stream):
+        metadatas = []
+        for block in self._resolved_external_objects:
+            metadata = {
+                "doi_or_unique_name": block[DataRegistryTarget.external_object, DataRegistryField.doi_or_unique_name],
+                MetadataKey.accessibility: block[DataRegistryTarget.storage_root, DataRegistryField.accessibility],
+                MetadataKey.version: block[DataRegistryTarget.external_object, DataRegistryField.version],
+                MetadataKey.verified_hash: block[DataRegistryTarget.storage_location, DataRegistryField.hash],
+                MetadataKey.extension: Path(block["output_filename"]).suffix[1:],
+                MetadataKey.filename: block["output_filename"],
+                MetadataKey.component: block[DataRegistryTarget.object_component, DataRegistryField.name],
+            }
+            metadatas.append(metadata)
+        if metadatas:
+            yaml.safe_dump(metadatas, stream)
+
+    def _download(self):
+        for block in itertools.chain(self._resolved_data_products, self._resolved_external_objects):
+            accessibility = block[DataRegistryTarget.storage_root, DataRegistryField.accessibility]
+            if accessibility == 0:  # public
+                source_uri = block[DataRegistryTarget.storage_root, DataRegistryField.root]
+                source_path = block[DataRegistryTarget.storage_location, DataRegistryField.path]
+                output_path = Path(block["full_output_filename"])
+                output_path.parent.mkdir(parents=True, exist_ok=True)
+
+                logger.info(f"Downloading public data from uri: {source_uri}, path: {source_path} to {output_path}")
+
+                protocol = urllib.parse.urlsplit(source_uri).scheme
+                fs, source_path = get_remote_filesystem_and_path(protocol, source_uri, source_path)
+                kwargs = {}
+                if not isinstance(fs, SFTPFileSystem):
+                    if fs.isdir(source_path):
+                        kwargs["recursive"] = True
+                    kwargs["block_size"] = 0
+                    fs.get(source_path, output_path.as_posix(), **kwargs)
+            else:
+                logger.info(f"Data is not public, skipping download")
+
+    def resolve(self):
+        """
+        Resolves all registered data products and external objects to their expanded registry data
+        """
+        logger.info(f"Resolving {len(self._data_products)} data product references")
+        while self._data_products:
+            block = self._data_products.pop()
```

Comment:
Again really a triviality, and maybe I just didn't quite get it, but your `_resolve_*` methods seem to iterate over their input as a list, but here you seem to be packaging them up individually in a list and then passing each one through (where the output list may be larger than the input list). Could you instead do it with a chain of iterators, something like this:
```python
def f(object_in):
    yield object_in + 1
    yield object_in + 2

def g(object_in):
    yield object_in * 2

def generate(f, iterator):
    for value in iterator:
        yield from f(value)

def process(functions, iterator):
    for function in functions:
        iterator = generate(function, iterator)
    return iterator

print(list(process((f, g), [1, 2, 3])))
print(list(process((g, f), [1, 2, 3])))
```